File: src/TqkLibrary.AudioPlayer.XAudio2.Native/XAudio2MasterVoice.cpp
#include "pch.h"
#include "XAudio2MasterVoice.h"

XAudio2MasterVoice::XAudio2MasterVoice(const XAudio2Engine* engine) {
	assert(engine);

	this->_xaudio2 = engine->_xaudio2;
	assert(this->_xaudio2);
}

XAudio2MasterVoice::~XAudio2MasterVoice() {
	if (this->_masterVoice)
	{
		this->_masterVoice->DestroyVoice();
		this->_masterVoice = nullptr;
	}
}

BOOL XAudio2MasterVoice::Init(int nb_channels, int sample_rate) {
	SetLastError(0);
	if (!this->_xaudio2)
		return FALSE;

	if (this->_masterVoice)
		return TRUE;

	HRESULT hr;
	hr = this->_xaudio2->CreateMasteringVoice(
		&this->_masterVoice,
		nb_channels,
		sample_rate
	); 
	SetLastError(hr);

	return SUCCEEDED(hr);
}

BOOL XAudio2MasterVoice::SetVolume(FLOAT volume) {
	SetLastError(0);
	if (!this->_masterVoice)
		return FALSE;

	HRESULT hr;
	hr = this->_masterVoice->SetVolume(volume);
	SetLastError(hr);
	return SUCCEEDED(hr);
}

VOID XAudio2MasterVoice::GetVolume(FLOAT* pVolume) {
	SetLastError(0);
	if (!this->_masterVoice)
		return;

	this->_masterVoice->GetVolume(pVolume);
}

BOOL XAudio2MasterVoice::SetChannelVolumes(UINT32 channels, const FLOAT* pVolume) {
	SetLastError(0);
	if (!this->_masterVoice)
		return FALSE;

	HRESULT hr;
	hr = this->_masterVoice->SetChannelVolumes(channels, pVolume);
	SetLastError(hr);
	return SUCCEEDED(hr);
}

VOID XAudio2MasterVoice::GetChannelVolumes(UINT32 channels, FLOAT* pVolume) {
	SetLastError(0);
	if (!this->_masterVoice)
		return;

	this->_masterVoice->GetChannelVolumes(channels, pVolume);
}

File: src/TqkLibrary.AudioPlayer.XAudio2/XAudio2MasterVoice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TqkLibrary.AudioPlayer.XAudio2
{
    public class XAudio2MasterVoice : IDisposable
    {
        readonly XAudio2Engine _engine;
        IntPtr _pointer = IntPtr.Zero;
        internal IntPtr Pointer { get { return _pointer; } }
        public float Volume
        {
            get
            {
                float volume = float.NaN;
                NativeWrapper.XAudio2MasterVoice_GetVolume(_pointer, ref volume);
                return volume;
            }
        }
        internal XAudio2MasterVoice(XAudio2Engine engine, IntPtr pAVFrame)
        {
            this._engine = engine ?? throw new ArgumentNullException(nameof(engine));
            _pointer = NativeWrapper.XAudio2MasterVoice_Alloc_AVFrame(engine.Pointer, pAVFrame);
            if (_pointer == IntPtr.Zero)
                throw new ApplicationException($"Create and load {nameof(XAudio2MasterVoice)} failed (last error : {NativeWrapper.GetLastError()})");
        }
        internal XAudio2MasterVoice(XAudio2Engine engine, int nb_channels, int sample_rate)
        {
            this._engine = engine ?? throw new ArgumentNullException(nameof(engine));
            _pointer = NativeWrapper.XAudio2MasterVoice_Alloc(engine.Pointer, nb_channels, sample_rate);
            if (_pointer == IntPtr.Zero)
                throw new ApplicationException($"Create and load {nameof(XAudio2MasterVoice)} failed (last error : {NativeWrapper.GetLastError()})");
        }
        ~XAudio2MasterVoice()
        {
            Dispose(false);
        }
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        void Dispose(bool disposing)
        {
            if (_pointer != IntPtr.Zero)
            {
                NativeWrapper.XAudio2MasterVoice_Free(ref _pointer);
            }
        }

        public bool SetVolume(float volume)
        {
            return NativeWrapper.XAudio2MasterVoice_SetVolume(_pointer, volume);
        }

        public bool SetChannelVolumes(float[] volumes)
        {
            return NativeWrapper.XAudio2MasterVoice_SetChannelVolumes(_pointer, (UInt32)volumes.Length, volumes);
        }
        public float[] GetChannelVolumes()
        {
            float[] volumes = Enumerable.Repeat<float>(-1.0f, 32).ToArray();
            NativeWrapper.XAudio2MasterVoice_GetChannelVolumes(_pointer, (UInt32)volumes.Length, ref volumes);
            return volumes.Where(x => x >= 0.0f).ToArray();
        }

        public XAudio2SourceVoice CreateSourceVoice(IntPtr pAVFrame)
        {
            return new XAudio2SourceVoice(this, pAVFrame);
        }
    }
}
